tee-worker/omni-executor/rpc-server/src/methods/pumpx/sign_limit_order.rs

```diff
@@ -0,0 +1,147 @@
+//
+// Litentry is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+use executor_primitives::utils::hex::FromHexPrefixed;
+use heima_primitives::Address32;
+use heima_primitives::IntentId;
+use rsa::pkcs1::DecodeRsaPrivateKey;
+use rsa::pkcs1::EncodeRsaPublicKey;
+
+use crate::error_code::get_native_task_error_code;
+use crate::error_code::AUTH_VERIFICATION_FAILED_CODE;
+use crate::server::RpcContext;
+use crate::ErrorCode;
+use ethers::types::Bytes;
+use executor_core::native_task::NativeTask;
+use executor_core::native_task::NativeTaskWrapper;
+use executor_core::native_task::PumpxChainId;
+use executor_core::native_task::PumxWalletIndex;
+use executor_crypto::jwt;
+use executor_primitives::OmniAuth;
+use heima_authentication::auth_token::AuthTokenClaims;
+use heima_primitives::Identity;
+use jsonrpsee::types::ErrorObject;
+use jsonrpsee::RpcModule;
+use log::error;
+use native_task_handler::{NativeTaskError, NativeTaskOk, NativeTaskResponse};
+use parity_scale_codec::Decode;
+use rsa::RsaPrivateKey;
+use serde::Deserialize;
+use serde::Serialize;
+use tokio::sync::oneshot;
+
+#[derive(Debug, Deserialize)]
+pub struct SignLimitOrderParams {
+	pub intent_id: IntentId,
+	pub order_id: u32,
+	pub chain_id: PumpxChainId,
+	pub wallet_index: PumxWalletIndex,
+	pub unsigned_tx: Vec<Bytes>,
+	pub auth_token: String,
+}
+
+#[derive(Serialize, Clone)]
+pub struct SignLimitOrderResponse {
+	pub intent_id: IntentId,
+	pub order_id: u32,
+	pub chain_id: PumpxChainId,
+	pub signed_tx: Vec<Bytes>,
+}
+
+pub fn register_sign_limit_order_params(module: &mut RpcModule<RpcContext>) {
+	module
+		.register_async_method("pumpx_signLimitOrder", |params, ctx, _| async move {
+			let internal_error: ErrorObject = ErrorCode::InternalError.into();
+
+			let params: SignLimitOrderParams = params.parse::<SignLimitOrderParams>()?;
+			let private_key = RsaPrivateKey::from_pkcs1_der(&ctx.jwt_rsa_private_key)
+				.map_err(|_| internal_error.clone())?;
+			let public_key =
+				private_key.to_public_key().to_pkcs1_der().map_err(|_| internal_error.clone())?;
+			// this validates jwt
+
+			let Ok(token) =
+				jwt::decode::<AuthTokenClaims>(&params.auth_token, public_key.as_bytes())
+			else {
+				return Err(ErrorCode::ServerError(AUTH_VERIFICATION_FAILED_CODE).into());
+			};
+			if token.sub != "access" {
+				return Err(ErrorCode::ServerError(AUTH_VERIFICATION_FAILED_CODE).into());
+			}
```

Comment:
Thank you

Shall we make a per task fn (like `expected_auth_token_type()` and add it as parameter to `verify_auth`) as it should be verified in most RPC methods?

Reply:
is it possible to reuse and keep alignment with other rpc calls?
Something like:
```
let wrapper: NativeTaskWrapper<NativeTask> = params.into();

if wrapper.task.require_auth() && verify_auth(ctx.clone(), &wrapper).await.is_err() {
return Err(ErrorCode::ServerError(AUTH_VERIFICATION_FAILED_CODE).into());
}
```

Reply:
> is it possible to reuse and keep alignment with other rpc calls? Something like:
> 
> ```
> let wrapper: NativeTaskWrapper<NativeTask> = params.into();
> 
> if wrapper.task.require_auth() && verify_auth(ctx.clone(), &wrapper).await.is_err() {
> return Err(ErrorCode::ServerError(AUTH_VERIFICATION_FAILED_CODE).into());
> }
> ```

Yes but that would require some time for rework.

Reply:
Just copy/paste and change fields? can it help?
Like:
```
impl From<RequestJwtParams> for NativeTaskWrapper<NativeTask> {
	fn from(p: RequestJwtParams) -> Self {
		Self {
			task: NativeTask::PumpxRequestJwt(
				Identity::from_web2_account(p.user_email.as_str(), Web2IdentityType::Email),
				p.invite_code,
				p.google_code,
				p.language,
			),
			nonce: None,
			auth: Some(OmniAuth::Email(p.email_code)),
		}
	}
}
```

Reply:
Nope, simple copy paste doesn't help in this case 😂

Reply:
The main problem is that `verify_auth` function derives omni_account from sender `Identity`.
In case of JWT `type: access` `sub` field already contains omni_account.